keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,440 @@
+some workload, a clear error will be shown.
+
+Please note the last sub-section here is a table with the summary of the changes
+proposed on each phase.
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volume work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and works only when no volume or any
+combination of the following volume types are used by the pod:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
+pods.
+
+The mapping length will be 64k, mapping the range 0-64k to the pod. This wide
+range makes sure most workloads will work fine.
+
+The mapping will be chosen by the kubelet, using a simple algorithm to give
+different pods in this category ("without" volumes) a non-overlapping mapping.
+Giving non-overlapping mappings generates the best isolation for pods.
+
+Furthermore, the node UID space of 2^32 can hold up to 64k pods each with a
+mapping length of 64k. This imposes a limit of 64k pods per node, but that is
+not an issue we will hit in practice for a long time, if ever (today we run 110
+pods per node by default).
+
+#### Phase 2: pods with volumes
+
+This phase makes user namespaces work in pods with volumes too. This is
+activated via the bool `pod.spec.HostUsers` too and pods fall into this mode if
+some other volume type than the ones listed for phase 1 is used. IOW, when phase
+2 is implemented, pods that use volume types not supported in phase 1, fall into
+the phase 2.
+
+All pods in this mode will use _the same_ mapping, chosen by the kubelet, with a
+length 64k, and mapping the range 0-64k too. IOW, each pod will have its own user
+namespace, but they will map to _the same_ UIDs/GIDs in the host.
+
+Using the same mapping allows for pods to share files and mitigates all the
+listed vulnerabilities (as the host is protected from the container). It is also
+a trivial next-step to take, given that we have phase 1 implemented: just return
+the same mapping if the pod has other volumes.
+
+#### Phase 3: pod to pod isolation
+
+This phase will provide more isolation between pods that use volumes (as in
+phase 2) and requires another opt-in field:
+`pod.spec.securityContext.pod2podIsolation`.
+
+This phase will try to not share the same mapping for all pods with volumes, as
+phase 2 does, but to achieve it some trade off needs to be made.
+
+One idea is to give different mappings to pods in different k8s namespaces or
+that use a different service account. This needs to be explored in further
+detail, but will probably impose limits to which workloads can run this (we need
+to expose a shorter mapping, less than 64k).
+
+Another idea is to use id mapped mounts. This probably needs changes to the
+OCI runtime-spec, only works with certain filesystems and kernels that may take
+too long for some users to get (like managed services).
```

Comment:
not sure if it is worth noticing, but I was playing with idmapped mounts to see how difficult would it be to support them for phase 3 and in crun I've implemented them using just a new option (so no changes to the OCI specs): https://github.com/containers/crun/pull/780

What do you think?

Reply:
Included it, thanks!